File: ValidateBinarySearchTree.cpp
#include <iostream>
using namespace std;

class Solution
{
public:
    bool isValidBST(TreeNode *root)
    {
        if (root == NULL)
            return false;
        if (root->left == NULL && root->right == NULL)
            return true;
        if (root->left != NULL && root->right == NULL)
            return isValidBST(root->left);
        if (root->right == NULL && root->right != NULL)
            return isValidBST(root->right);
        if (root->)
        
    }
};

int main()
{
    return 0;
}

File: JavaSolution/src/ValidateBinarySearchTree.java
import java.util.ArrayList;
import java.util.List;

/**
 * Created by xyunpeng on 7/8/15.
 * TODO: Need fully understand
 */
public class ValidateBinarySearchTree {
    public boolean isValidBST(TreeNode root) {
        return helper(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }

    public boolean helper(TreeNode t, long min, long max) {
        if (t == null)
            return true;
        if (t.val <= min || t.val >= max)
            return false;
        return helper(t.left, min, t.val) && helper(t.right, t.val, max);
    }
}
